# Annotations

## Setup

```{r, message = F, warning = F}
library(tidyverse)
```

## Set variables

```{r}
SIGS = list("/hps/nobackup/birney/users/ian/MIKK_HMM/sig_snps/hdrr/dist_angle/dist_angle/15/5000/0.8/invnorm/open_field_sigs.csv",
            "/hps/nobackup/birney/users/ian/MIKK_HMM/sig_snps/hdrr/dist_angle/dist_angle/15/5000/0.8/invnorm/novel_object_sigs.csv")
names(SIGS) = c("open_field", "novel_object")
VEP = list(here::here("results/annotations/hdrr/dist_angle/dist_angle/15/5000/0.8/invnorm/open_field_vep_out.txt"),
           here::here("results/annotations/hdrr/dist_angle/dist_angle/15/5000/0.8/invnorm/novel_object_vep_out.txt"))
names(VEP) = c("open_field", "novel_object")
```

## Read in files

```{r, message = F, warning = F}
sigs = purrr::map_dfr(SIGS, ~readr::read_csv(.x,
                                             col_types = c("ccdddcccccdddd")),
                      .id = "ASSAY")

vep = purrr::map_dfr(VEP, ~readr::read_tsv(.x, comment = "##") %>% 
                      dplyr::rename("Uploaded_variation" = "#Uploaded_variation"),
                    .id = "ASSAY") %>% 
  # split `Uploaded_variation` into CHROM, POS, REF, ALT
  tidyr::separate(col = "Uploaded_variation",
                  into = c("CHROM", "POS", "ALLELES"),
                  sep = "_") %>% 
  tidyr::separate(col = "ALLELES",
                  into = c("REF", "ALT"),
                  sep = "/") %>% 
  # conver CHROM to integer
  dplyr::mutate(CHROM = as.integer(CHROM),
                POS = as.integer(POS))
```

## Within each assay, are there any SNPs that are shared across DGE/SGE?

```{r}
split_sigs = sigs %>% 
  split(., f = .$ASSAY) %>% 
  purrr::map(., ~split(.x, f = .x$DGE_SGE))

# Any shared for open field?
which(split_sigs$open_field$dge %in% split_sigs$open_field$sge)

# Any shared for novel object?
which(split_sigs$novel_object$dge %in% split_sigs$novel_object$sge)
```

## Greedily filter SNPs

```{r}
WINDOW_LEN = 1e5

# Function for greedy filtering
greedy_drop = function(df){
  # sort `df` by p-value
  df = df %>% 
    dplyr::arrange(p)
  # add first row (lowest p-value) to output
  out = df[1, ]
  # for all other rows...
  if (nrow(df) > 1){
    for (i in 2:nrow(df)){
      # if its absolute bp distance from any other kept SNP is greater than the window length
      if (min(abs(df[[i,"POS"]] - out$POS)) > WINDOW_LEN){
        # add that row to the output
        out = out %>% 
          tibble::add_row(df[i, ])
      }
    }    
  }
  return(out)
}

# Apply function to each combination
greedy_kept = split(sigs, ~ASSAY + DGE_SGE + STATE + CHROM) %>% 
  purrr::map(~ .x %>% 
               dplyr::arrange(p)) %>% 
  # drop empty data frames
  purrr::discard(~nrow(.x) == 0) %>% 
  purrr::map_dfr(~greedy_drop(.x))
  

```

## Join `sigs` to `vep`

```{r}
sigs_vep = sigs %>% 
  dplyr::full_join(vep,
                    by = c("ASSAY", "CHROM", "POS", "REF", "ALT")) %>% 
  # remove unnecessary columns
  dplyr::select(-c(cDNA_position, CDS_position, Protein_position, Amino_acids, Codons, Existing_variation))

```


## Get gene names from `biomaRt`

```{r}
olat_mart = biomaRt::useEnsembl(biomart = "ensembl", dataset = "olatipes_gene_ensembl", mirror = "uswest")

olat_attr = biomaRt::listAttributes(olat_mart)

olat_genes = biomaRt::getBM(attributes = c("chromosome_name",
                                           "start_position",
                                           "end_position",
                                           "ensembl_gene_id",
                                           "hgnc_symbol",
                                           "description"),
                            mart = olat_mart) 

# Convert HdrR genes to GenomicRanges
olat_genes_r = olat_genes %>% 
  GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "chromosome_name",
                                          start.field = "start_position",
                                          end.field = "end_position")


# SNPs

## convert SNP hits to genomic ranges
snp_loc_r = sigs_vep %>% 
  GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "CHROM",
                                          start.field = "POS",
                                          end.field = "POS",
                                          ignore.strand = T)


## find overlaps
olaps_snps = GenomicRanges::findOverlaps(snp_loc_r, olat_genes_r)

# Pull out data frame of hits

snp_hits_all = dplyr::bind_cols(sigs_vep %>% 
                                  dplyr::slice(olaps_snps@from),
                                olat_genes[olaps_snps@to, ])

```

## View results

```{r}
# Counts for different consequences
snp_hits_all %>% 
  dplyr::count(Consequence) %>% 
  DT::datatable(.)

# Pull out missense variants
snp_hits_all %>% 
  dplyr::filter(Consequence == "missense_variant") %>% 
  DT::datatable(.)
```

